src/render/draw_heatmap.js

```diff
@@ -0,0 +1,149 @@
+        // to use complex tile masking here because the change between zoom levels is subtle,
+        // so it's fine to simply render the parent until all its 4 children are loaded
+        if (sourceCache.hasRenderableParent(coord)) continue;
+
+        const tile = sourceCache.getTile(coord);
+        const bucket: ?HeatmapBucket = (tile.getBucket(layer): any);
+        if (!bucket) continue;
+
+        const programConfiguration = bucket.programConfigurations.get(layer.id);
+        const program = painter.useProgram('heatmap', programConfiguration);
+        const {zoom} = painter.transform;
+        programConfiguration.setUniforms(gl, program, layer, {zoom});
+        gl.uniform1f(program.uniforms.u_radius, layer.getPaintValue('heatmap-radius', {zoom}));
+
+        gl.uniform1f(program.uniforms.u_extrude_scale, pixelsToTileUnits(tile, 1, zoom));
+
+        gl.uniform1f(program.uniforms.u_intensity, layer.getPaintValue('heatmap-intensity', {zoom}));
+        gl.uniformMatrix4fv(program.uniforms.u_matrix, false, coord.posMatrix);
+
+        program.draw(
+            gl,
+            gl.TRIANGLES,
+            layer.id,
+            bucket.layoutVertexBuffer,
+            bucket.indexBuffer,
+            bucket.segments,
+            programConfiguration);
+    }
+
+    renderTextureToMap(gl, painter, layer);
+}
+
+function renderToTexture(gl, painter, layer) {
+    gl.activeTexture(gl.TEXTURE1);
+
+    // Use a 4x downscaled screen texture for better performance
+    gl.viewport(0, 0, painter.width / 4, painter.height / 4);
+
+    let texture = layer.heatmapTexture;
+    let fbo = layer.heatmapFbo;
+
+    if (!texture) {
+        texture = layer.heatmapTexture = gl.createTexture();
+        gl.bindTexture(gl.TEXTURE_2D, texture);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+
+        fbo = layer.heatmapFbo = gl.createFramebuffer();
+
+        bindTextureFramebuffer(gl, painter, texture, fbo);
+
+    } else {
+        gl.bindTexture(gl.TEXTURE_2D, texture);
+        gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+    }
+}
+
+function bindTextureFramebuffer(gl, painter, texture, fbo) {
+    // Use the higher precision half-float texture where available (producing much smoother looking heatmaps);
+    gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, painter.width / 4, painter.height / 4, 0, gl.RGBA,
+        painter.extTextureHalfFloat ? painter.extTextureHalfFloat.HALF_FLOAT_OES : gl.UNSIGNED_BYTE, null);
+
+    gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+    gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, texture, 0);
+
+    // If using half-float texture as a render target is not supported, fall back to a low precision texture
```

Comment:
Is there any way we can test unavailability of this extension? For VAOs, I think we have code that specifically prevents reporting that this extension is available so that we can run the test with it enabled and disabled.

Reply:
Unfortunately the `gl` module which we use for tests doesn't seem to support the extension, so we're forced to only test the "fallback" state.

Also, this check is not about availability of the halfFloat extension, it's about being able to use halfFloat textures as a render target (which some GPUs simply don't support while they support the extension). So the only way to know that is to try, and then check the framebuffer attachment status (this is even officially documented somewhere).

Reply:
Found this thread on the matter https://groups.google.com/forum/#!topic/webgl-dev-list/ms21hnWk0tg